File: DesignPatterns/Behavioral/ChainOfResponsibility/NumberTypeClient.java
interface Chain {
  public abstract void setNext(Chain next);
  public abstract void process(Number request);
}

class Number {
  private int number;

  public Number(int number) {
    this.number = number;
  }

  public int getNumber(){
    return this.number;
  }
}

class NegativeProcessor implements Chain {

  private Chain next;

  public void setNext(Chain next) {
    this.next = next;
  }

  public void process(Number request) {
    if (request.getNumber() < 0) {
      System.out.println("NegativeProcessor : " + request.getNumber());
    } else {
      next.process(request);
    }
  }
}

class ZeroProcessor implements Chain {

  private Chain next;

  public void setNext(Chain next) {
    this.next = next;
  }

  public void process(Number request) {
    if(request.getNumber() == 0) {
      System.out.println("ZeroProcessor : " + request.getNumber());
    } else {
      next.process(request);
    }
  }
}

class PositiveProcessor implements Chain {

  private Chain next;

  public void setNext(Chain next) {
    this.next = next;
  }

  public void process(Number request) {
    if(request.getNumber() > 0) {
      System.out.println("PositiveProcessor : " + request.getNumber());
    } else {
      next.process(request);
    }
  }
}


public class NumberTypeClient {
  public static void main(String[] args) {
    Chain c1 = new NegativeProcessor();
    Chain c2 = new ZeroProcessor();
    Chain c3 = new PositiveProcessor();

    c1.setNext(c2);
    c2.setNext(c3);

    c1.process(new Number(90));
    c1.process(new Number(-90));
    c1.process(new Number(0));
    c1.process(new Number(91));
  }
}

File: DesignPatterns/Behavioral/ChainOfResponsibility/Ruby/number_type_client.rb
class ChainProcessor
  attr_reader :next_handler

  def set_next_handler(next_handler)
    @next_handler = next_handler
  end

  def process(number)
    raise NotImplementedError
  end
end

class NegativeProcesssor < ChainProcessor
  def process(number)
    if number < 0
      puts "Handled by negative processor" 
    else
      @next_handler.process(number)
    end
  end
end

class PositiveProcessor < ChainProcessor
  def process(number)
    if number > 0
      puts "Handled by positive processor"
    else
      @next_handler.process(number)
    end
  end
end

class ZeroProcessor < ChainProcessor
  def process(number)
    if number == 0
      puts "Handled by zero processor"
    else
      @next_handler.process(number) 
    end
  end
end

np = NegativeProcesssor.new
pp = PositiveProcessor.new
zp = ZeroProcessor.new

np.set_next_handler(pp)
pp.set_next_handler(zp)
zp.set_next_handler(np)

np.process(90)
np.process(-90)
np.process(0)
